riscv/csrs.cc

```diff
@@ -504,9 +505,9 @@ bool misa_csr_t::unlogged_write(const reg_t val) noexcept {
   if (!(val & (1L << ('C' - 'A'))) && (state->pc & 2))
     return false;
 
-  const bool val_supports_f = val & (1L << ('F' - 'A'));
-  const reg_t val_without_d = val & ~(1L << ('D' - 'A'));
-  const reg_t adjusted_val = val_supports_f ? val : val_without_d;
+  reg_t adjusted_val = val;
+  adjusted_val = (val & (1L << ('F' - 'A'))) ? adjusted_val : (val & ~(1L << ('D' - 'A')));
+  adjusted_val = (val & (1L << ('D' - 'A'))) ? adjusted_val : (val & ~(1L << ('Q' - 'A')));
```

Comment:
The key thing here is a dependency between F/D and Q/D. Can we find a good way to express this succinctly? For example the code could read something like:
```
adjusted_val = dependency(adjusted_val, 'D', 'F');
adjusted_val = dependency(adjusted_val, 'Q', 'D');
```
And if we could make it even more functional by removing the mutable variable that would be even better.

Then, make this change first as a refactor of the existing F/D bits, in its own commit, before adding the new functionality in later commits.